dune/iga/nurbstrimfunctionality.hh

```diff
@@ -0,0 +1,624 @@
+
+    Clipper2Lib::RectD elementRect{corners[0][0], corners[1][1], corners[1][0], corners[3][1]};
+
+    // Clip with RectClip (more efficient than normal Intersect Clip, works because our ParameterSpace Grid
+    // has always rectangular elements
+    Clipper2Lib::PathsD clippedEdges = Clipper2Lib::RectClip(elementRect, clip, clipperPrecision);
+
+    // If the clippedEdges are empty, this means the element is outside of the clip -> empty
+    // And if the area of the clip is equal to the area of the element -> full
+    if (clippedEdges.empty())
+      trimFlag = ElementTrimFlag::empty;
+    else {
+      if (FloatCmp::eq(Clipper2Lib::Area(clippedEdges), element.geometry().volume(), tolerance))
+        trimFlag = ElementTrimFlag::full;
+      else
+        trimFlag = ElementTrimFlag::trimmed;
+    }
+
+    if (trimFlag != ElementTrimFlag::trimmed) return std::make_pair(trimFlag, std::nullopt);
+
+    // If the code is here → Elements are trimmed
+    assert(clippedEdges.size() == 1);
+    clippedEdges.front() = Clipper2Lib::TrimCollinear(clippedEdges.front(), clipperPrecision);
+
+    // Create a local pointMap to store the intersection points in regard to their corresponding edges or nodes
+    // Node Intersection Points are stored with indices 10 - 13 & edges with 0 - 3
+    IntersectionPointMap pointMap;
+
+    for (auto& point : clippedEdges[0]) {
+      // Check if point is on any Edge
+      auto [isOnAnyEdge, edgeThePointIsOn] = pointOnAnyEdge(point, elementEdges);
+      auto [isOnAnyNode, nodeThePointIsOn] = pointOnAnyVertex(point, corners);
+
+      // If is onAnyNode is true, then isOnAnyEdge has to be also true
+      if (isOnAnyNode)
+        pointMap[nodeThePointIsOn + 10].emplace_back(point);
+      else if (isOnAnyEdge)
+        pointMap[edgeThePointIsOn].emplace_back(point);
+    }
+
+    // This keeps track of how many intersection Points there are at a given edge
+    std::array<int, 4> edgeCounter{0, 0, 0, 0};
+    std::array<int, 4> nodeCounter{0, 0, 0, 0};
+
+    for (const auto& [key, value] : pointMap) {
+      if (key < 4)
+        edgeCounter[key] = static_cast<int>(value.size());
+      else
+        nodeCounter[key - 10] = static_cast<int>(value.size());
+    }
+
+    // Make a shared Pointer out of pointMap
+    IntersectionPointMapPtr pointMapPtr = std::make_shared<IntersectionPointMap>(pointMap);
+
+    // Sort IntersectionPoints counter-clockwise
+    for (int i = 0; i < 4; ++i)
+      sortIntersectionPoints(pointMapPtr, i);
+
+    return std::make_pair(trimFlag,
+                          std::make_optional<ClippingResult>({std::move(pointMapPtr), edgeCounter, nodeCounter}));
+  }
+
+  bool areEdgesParallel(CurveGeometry& curve1, CurveGeometry& curve2) {
+    if (curve1.degree()[0] > 1 || curve2.degree()[0] > 1) return false;
+
+    auto [p1, p2] = curveStartEndPoint(curve1);
+    auto [p3, p4] = curveStartEndPoint(curve2);
+
+    // Compute direction vectors of both curves
```

Comment:
Instead of curve you can use line or straight curve to make it clear that you expect straight curve at that point.

Reply:
Done